### PR TITLE
V2 Review and Testing

### PR DESCRIPTION
**Pull request by aferditamuriqi**
@kristojorg I am finalizing my testing, reviewing and fixing issues where they come up, as well as adding documentation. over the next ~2 weeks i will be adding to the draft PR and then once finalized and you get a change to review, i am ready to merge up to develop. targeting to finalize it before the holidays :) 

@pszujewski @Toxiapo @hokei please feel free to share your feedback as well. :) 

as usual i am combining commits based on feature/fixes/cleanup etc. for easier review. 

included in this PR also this https://d-i-t-a.github.io/R2D2BC/classes/D2Reader.html (keep in mind this is WIP as i keep adding comments within the code) 

fixes https://github.com/d-i-t-a/R2D2BC/issues/294 (isScrollMode sometimes is called too early)
fixes https://github.com/d-i-t-a/R2D2BC/issues/293 (Paginated view text truncation in v2)
fixes https://github.com/d-i-t-a/R2D2BC/issues/215 (Make findRequiredElement generic V2)
fixes https://github.com/d-i-t-a/R2D2BC/issues/214 where possible (Make methods into getters wherever possible on D2Reader V2)
fixes https://github.com/d-i-t-a/R2D2BC/issues/206 (Enable strictNullChecks V2)
fixes https://github.com/d-i-t-a/R2D2BC/issues/205 for rights (Distinct module input options from internal settings V2)

### CODE REVIEW

**Comment by aferditamuriqi** [comment]
@Toxiapo In this PR are quite a few scroll-mode fixes, also includes the fix for the scroll height. feel free to review this, i am ready to get this merged. i also have the strict null check PR ready, which is based on this branch. @kristojorg if you get a chance to review that PR, that would be very much appreciated :) https://github.com/d-i-t-a/R2D2BC/pull/316

File: examples/react/tsconfig.json [hunk 1]
```diff
@@ -34,9 +34,9 @@
     "sourceMap": true,
     "declaration": true,
     "preserveConstEnums": true,
-    "strictNullChecks": false,
+    "strictNullChecks": true,
```
Comment (kristojorg) [inline examples/react/tsconfig.json:37]: 🎉

File: src/index.ts [hunk 2]
```diff
@@ -18,10 +18,8 @@
...
 import D2Reader from "./reader";
 
+/** R2D2BC Reader */
 export default D2Reader;
 
-// for interop with <script> based usage
-export const build = D2Reader.build;
-
-export { Link } from "./model/Link";
-export { Locator, ReadingPosition } from "./model/Locator";
```
Comment (kristojorg) [inline src/index.ts]: I think you will want to export these items as they are useful to consumers of the library. I would recommend people don't import from `/dist` directories, but instead that everything that they might need can be imported from the root via `@d-i-t-a/reader` directly.

File: src/modules/BookmarkModule.ts [hunk 3]
```diff
@@ -224,194 +223,179 @@ export default class BookmarkModule implements ReaderModule {
...
+              bookmark = result;
+            }
+            if (IS_DEV) console.log(bookmark);
+            let saved = this.annotator.saveBookmark(bookmark);
 
-          if (IS_DEV) {
-            console.log("Bookmark added " + JSON.stringify(saved));
+            if (IS_DEV) {
+              console.log("Bookmark added " + JSON.stringify(saved));
+            }
```
Comment (kristojorg) [inline src/modules/BookmarkModule.ts:298]: Just a thought for the future: You could add a `log()` function that incorporates the `IS_DEV` check within itself

File: src/modules/sampleread/SampleReadEventHandler.ts [hunk 4]
```diff
@@ -64,31 +67,28 @@ export default class SampleReadEventHandler {
...
+
+    let opts =
+      Object.defineProperty &&
+      Object.defineProperty({}, "passive", {
+        // eslint-disable-next-line getter-return
+        get: function () {
+          supportsPassive = true;
+        },
+      });
+    window.addEventListener("test", function () {}, opts);
```
Comment (kristojorg) [inline src/modules/sampleread/SampleReadEventHandler.ts:79]: Is this meant to be here?

**Review by kristojorg — APPROVED** [review]
Looks good to me! I admittedly didn't have time to read line-by-line, but I scanned through every file : ). These are some exciting updates!

File: src/modules/sampleread/SampleReadEventHandler.ts (hunk 4 above)
Comment (aferditamuriqi) [reply src/modules/sampleread/SampleReadEventHandler.ts:79]: yes it is, it's a refactor from the lines above, after turning strict null check on, some of the syntax from before was acting up

File: src/index.ts (hunk 2 above)
Comment (aferditamuriqi) [reply src/index.ts]: okey, sure i will revert this then.

File: src/modules/BookmarkModule.ts (hunk 3 above)
Comment (aferditamuriqi) [reply src/modules/BookmarkModule.ts:298]: That's a good point, i'll add it to a todo, won't do this now though, just to not delay much longer.

File: src/index.ts (hunk 2 above)
Comment (aferditamuriqi) [reply src/index.ts]: reverted.

**Pull request merged**